src/components/organisms/NewProfileModal/UserProfileModal/UserProfileModalContent/UserProfileModalContent.tsx

```diff
@@ -0,0 +1,244 @@
+
+  const {
+    isShown: isSubmitting,
+    show: startSubmitting,
+    hide: finishSubmitting,
+  } = isSubmittingState;
+
+  const defaultValues = useProfileModalFormDefaultValues(
+    user,
+    questions,
+    answers
+  );
+
+  const {
+    register,
+    reset,
+    errors,
+    setError,
+    clearError,
+    watch,
+    handleSubmit,
+    getValues,
+    setValue,
+    control,
+  } = useForm<UserProfileModalFormData>({
+    mode: "onBlur",
+    reValidateMode: "onChange",
+    validateCriteriaMode: "all",
+    defaultValues,
+  });
+
+  const {
+    fields,
+    append: addLink,
+    remove: removeLink,
+  } = useFieldArray<ProfileLink>({
+    control,
+    name: formProp("profileLinks"),
+  });
+
+  const links = fields as WithId<ProfileLink>[];
+
+  const cancelEditing = useCallback(() => {
+    turnOffEditMode();
+    reset();
+  }, [reset, turnOffEditMode]);
+
+  const addLinkHandler = useCallback(() => {
+    addLink({ url: "", title: "" });
+  }, [addLink]);
+
+  const checkOldPassword = useCallback(
+    async (password: string) => {
+      if (!firebaseUserEmail) return;
+      try {
+        await firebase
+          .auth()
+          .signInWithEmailAndPassword(firebaseUserEmail, password);
+        return true;
+      } catch {
+        return false;
+      }
+    },
+    [firebaseUserEmail, firebase]
+  );
+
+  const setLinkTitle = useCallback(
+    (index: number, title: string) => {
+      setValue(
+        `${formProp("profileLinks")}[${index}].${propName<ProfileLink>(
+          "title"
+        )}`,
+        title
+      );
+    },
+    [setValue]
+  );
+
+  const onSubmit: OnSubmit<UserProfileModalFormData> = useCallback(
+    async (data) => {
+      if (!firebaseUser) return;
+
+      startSubmitting();
+      try {
+        const passwordsNotEmpty = arePasswordsNotEmpty(data);
+        if (passwordsNotEmpty) {
+          if (!(await checkOldPassword(data.oldPassword))) {
+            setError(formProp("oldPassword"), "validate", "Incorrect password");
+            return;
+          } else {
+            clearError(formProp("oldPassword"));
+            await firebaseUser.updatePassword(data.confirmNewPassword);
+          }
+        }
+
+        const fieldsToCheck = [
+          formProp("partyName"),
+          formProp("pictureUrl"),
+          ...questions.map((x) => x.name),
+        ];
+
+        const changedFields = fieldsToCheck.filter(
+          (field) =>
+            !isEqual(
+              defaultValues[field as keyof typeof defaultValues],
+              data[field as keyof typeof data]
+            )
+        ) as (keyof typeof data)[];
+
+        if (changedFields.length > 0) {
+          await updateUserProfile(firebaseUser.uid, pick(data, changedFields));
+        }
+
+        if (!isEqual(user.profileLinks, data.profileLinks)) {
+          await updateProfileLinks({
+            profileLinks: data.profileLinks ?? [],
+            userId: user.id,
+          });
```

Comment:
This logic looks really complicated. We should definitely try to simplify it. Have you checked if useForm has proper API for dirty fields?

Reply:
Indeed it does, thank you! Fixed in https://github.com/sparkletown/sparkle/pull/1934/commits/0c31a66aa878f5517a180eb9a170e091bf5e2ae5